lib/Models/RegionProvider.js

```diff
@@ -0,0 +1,405 @@
+        }));
+    }
+    return when.all(p).yield(true);
+};
+
+
+function applyReplacements(s, replacements) {
+// replacements: [ [ regex, replacement], ... ] 
+// replacement can contain '$1' etc.
+    var r = s.trim().toLowerCase(); // just in case...
+
+    if (replacements === undefined || replacements.length === 0) {
+        return r;
+    }
+
+    var cachekey = s + '%' + replacements.reduce(function(key, item) { 
+        return key + item[0] + '%' + (item[1] ? item[1] : '') + '%'; 
+        }, '');
+    if (applyReplacements.cache[cachekey] !== undefined) {
+        return applyReplacements.cache[cachekey];
+    }
+    
+    replacements.forEach(function(rep) {
+        r = r.replace(new RegExp(rep[0].toLowerCase(), 'gi'), rep[1].toLowerCase());
+    });
+    applyReplacements.cache[s] =r;
+    return r;
+}
+applyReplacements.cache = {};
+
+
+/* Checks if a given region code matches a given ID after applying regexs and additional disambiguation lookups */
+RegionProvider.prototype.codeMatchesRegionID = function(id, code, disambigId, disambigCode) { //, row, idnum, csvItem ) {
+    code = String(code).toLowerCase(); //## should be happening upstream somewhere
+    id = String(id).toLowerCase(); 
+    if (applyReplacements(id, this.serverReplacements) !== applyReplacements(code, this.dataReplacements)) {
+        return false; // failed match, even after best efforts
+    }
+    if (!defined(this.disambigColumns) || this.disambigColumns.length === 0) {
+        return true; // unambiguous successful match
+    }
+
+    if (defined(disambigId) && !defined(disambigCode)) {
+        // warn? we have a value to disambiguate against, but it's not provided
+        return true;
+    }
+
+    // Hooray, let's disambiguate. Our args look something like: Campbelltown, Campbelltown, New South Wales, NSW.
+
+    var dabMatched = this._getDisambigProvider().codeMatchesRegionID(disambigId, disambigCode);
+    if (!dabMatched) {
+        console.log ('"' + id + '" matched "' + code + '", but "' + disambigId + '" didn\'t match "' + disambigCode + '"');
+    }
+    return dabMatched;
+};
+
+/**
+ * Given a region code, try to find a region that matches it, using replacements, disambiguation and other wizardry.
+ * @param {String} code Code to search for.
+ * @returns {Number} Zero-based index in list of regions if successful, or -1.
+ */
+RegionProvider.prototype.findRegionIndex = function(code, disambigCode) { //, ids, rd, row) {
+    for (var i = 0; i < this.regions.length; i++) {
+        var dabId;
+        if (defined(this.disambigColumns) && defined(this.disambigColumns[0])) {
+            dabId =  this.regions[i][this.disambigColumns[0].regionProp];    
+        }
+        if (this.codeMatchesRegionID(this.regions[i].id, code, dabId, disambigCode)) {
+            return i;
+        }
+    }
+    return -1;
+};
+
+/**
+ *   Provides a list of values for each region, based on a provided dataset.
+ *
+ *   @param {Dataset} dataset
```

Comment:
Fixed. There seems to be a convention of variables called `dataset` referring to `DataTable` objects.